Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,505 @@
+  · rw [← rank_eq_one_iff_finrank_eq_one, rank_subsingleton]
+  · rw [(free_iff_linearEquiv.mp ‹_›).some.finrank_eq, finrank_self]
+
+theorem rank_eq_one [StrongRankCondition R] [Free R M] : Module.rank R M = 1 :=
+  rank_eq_one_iff_finrank_eq_one.mpr (finrank_eq_one R M)
+
+open TensorProduct (comm lid) in
+theorem toModuleEnd_bijective : Function.Bijective (toModuleEnd R (S := R) M) := by
+  have : toModuleEnd R (S := R) M = (lid R M).conj ∘ rTensorEquiv R R
+      (comm .. ≪≫ₗ linearEquiv R M) ∘ RingEquiv.moduleEndSelf R ∘ MulOpposite.opEquiv := by
+    ext; simp [LinearEquiv.conj, liftAux]
+  simpa [this] using MulOpposite.opEquiv.bijective
+
+instance : FaithfulSMul R M where
+  eq_of_smul_eq_smul {_ _} h := (toModuleEnd_bijective R M).injective <| LinearMap.ext h
+
+variable {R M N} in
+private theorem bijective_self_of_surjective (f : R →ₗ[R] M) (hf : Function.Surjective f) :
+    Function.Bijective f where
+  left {r₁ r₂} eq := smul_left_injective' (α := M) <| funext fun m ↦ by
+    obtain ⟨r, rfl⟩ := hf m
+    simp_rw [← map_smul, smul_eq_mul, mul_comm _ r, ← smul_eq_mul, map_smul, eq]
+  right := hf
+
+variable {R M N} in
+/- Not true if `surjective` is replaced by `injective`: any nonzero element in an invertible
+module over a domain generates a submodule isomorphic to the domain, which is not the whole
+module unless the module is free. -/
+theorem bijective_of_surjective [Module.Invertible R N] (f : M →ₗ[R] N)
+    (hf : Function.Surjective f) : Function.Bijective f := by
+  simpa [lTensor_bijective_iff] using bijective_self_of_surjective
+    (f.lTensor _ ∘ₗ (linearEquiv R M).symm.toLinearMap) (by simpa [lTensor_surjective_iff] using hf)
+
+section Algebra
+
+instance : Module.Invertible A (A ⊗[R] M) :=
+  .right (M := A ⊗[R] Dual R M) <| (AlgebraTensorModule.distribBaseChange ..).symm ≪≫ₗ
+    AlgebraTensorModule.congr (.refl A A) (linearEquiv R M) ≪≫ₗ AlgebraTensorModule.rid ..
+
+instance (L) [AddCommMonoid L] [Module R L] [Module A L] [IsScalarTower R A L]
+    [Module.Invertible A L] : Module.Invertible A (L ⊗[R] M) :=
+  .congr (AlgebraTensorModule.cancelBaseChange R A A L M)
+
+variable [FaithfulSMul R A] [Free A (A ⊗[R] M)]
+
+/-- An invertible `R`-module embeds into an `R`-algebra that `R` injects into,
+provided `A ⊗[R] M` is a free `A`-module. -/
+noncomputable def embAlgebra : M →ₗ[R] A :=
+  (free_iff_linearEquiv.mp ‹_›).some.restrictScalars R ∘ₗ
+    (Algebra.ofId R A).toLinearMap.rTensor M ∘ₗ (TensorProduct.lid R M).symm
+
+theorem embAlgebra_injective : Function.Injective (embAlgebra R M A) := by
+  simpa [embAlgebra] using
+    Flat.rTensor_preserves_injective_linearMap _ (FaithfulSMul.algebraMap_injective R A)
+
+/-- An invertible `R`-module as a `R`-submodule of an `R`-algebra. -/
+noncomputable def toSubmodule : Submodule R A := LinearMap.range (embAlgebra R M A)
+
+end Algebra
+
+end Invertible
+
+end Module
+
+section PicardGroup
+
+open CategoryTheory Module
+
+variable (R : Type u) [CommRing R]
```

Comment:
Isn't `R` already in scope?

Reply:
Nope, everything above is enclosed in the Module namespace ... (and applies to CommSemiring)